tests/backup/backup_test.go

```diff
@@ -1964,9 +1960,379 @@ var _ = Describe("{ShareBackupsAndClusterWithUser}", func() {
+					defer wg.Done()
+					defer func() { <-sem }()
+					err = CreateBackup(backupName, SourceClusterName, bkpLocationName, backupLocationUID, []string{bkpNamespaces[0]},
+						labelSelectors, orgID, clusterUid, "", "", "", "", ctx)
+					dash.VerifyFatal(err, nil, "Taking backup of application")
+				}(backupName)
+			}
+			wg.Wait()
+			log.Infof("List of backups - %v", backupNames)
+		})
+
+		Step("Sharing backup with multiple users with different access permission and different roles", func() {
+			log.InfoD("Sharing backup with multiple users with different access permission and different roles")
+			backupsToBeSharedWithUserOfSameRole := 3
+			for i := 0; i < numberOfUsers; i++ {
+				userIndex := i % backupsToBeSharedWithUserOfSameRole
+				switch userIndex {
+				case 0:
+					ShareBackup(backupNames[i], nil, []string{users[i]}, ViewOnlyAccess, ctx)
+					log.Infof(" Backup %s shared with user %s", backupNames[i], users[i])
+				case 1:
+					ShareBackup(backupNames[i], nil, []string{users[i]}, RestoreAccess, ctx)
+					log.Infof(" Backup %s shared with user %s", backupNames[i], users[i])
+				case 2:
+					ShareBackup(backupNames[i], nil, []string{users[i]}, FullAccess, ctx)
+					log.Infof(" Backup %s shared with user %s", backupNames[i], users[i])
+				default:
+					ShareBackup(backupNames[i], nil, []string{users[i]}, ViewOnlyAccess, ctx)
+					log.Infof(" Backup %s shared with user %s", backupNames[i], users[i])
+				}
+			}
+		})
+		Step("Validating the shared backup with user having RestoreAccess and  px-backup-app.admin role", func() {
+			log.InfoD("Validating if user with  RestoreAccess and px-backup-app.admin role can restore but not delete the backup ")
+			ctxNonAdmin, err := backup.GetNonAdminCtx(users[1], "Password1")
+			log.FailOnError(err, "Fetching user ctx")
+			userContexts = append(userContexts, ctxNonAdmin)
+			log.InfoD("Registering Source and Destination clusters from user context")
+			CreateSourceAndDestClusters(orgID, "", "", ctxNonAdmin)
+			// Try restore with user having RestoreAccess and it should pass
+			restoreName := fmt.Sprintf("%s-%s-%v", users[1], RestoreNamePrefix, time.Now().Unix())
+			userRestoreContext[ctxNonAdmin] = restoreName
+			err = CreateRestore(restoreName, backupNames[1], make(map[string]string), destinationClusterName, orgID, ctxNonAdmin)
+			dash.VerifyFatal(err, nil, "Verifying that restore is possible")
+			// Try to delete the backup with user having RestoreAccess,and it should not pass
+			backupUID, err := backupDriver.GetBackupUID(ctx, backupNames[1], orgID)
+			log.FailOnError(err, "Failed while trying to get backup UID for - %s", backupNames[1])
+			// Delete backup to confirm that the user has RestoreAccess,and it cannot delete backup
+			_, err = DeleteBackup(backupNames[1], backupUID, orgID, ctxNonAdmin)
+			dash.VerifyFatal(strings.Contains(err.Error(), "doesn't have permission to delete backup"), true, "Verifying backup deletion is not possible")
```

Comment:
This whole block can be converted into a function and re-used. All you have to pass to this function is context, backup name, access to be validated